### PR TITLE
torch_xla multi-device support

### PR DESCRIPTION
**Pull request by dskhudia**
Adds support for multi-device training using torch_xla. 

The resnet9 on cifar10 trains fine on 2 GPUs using torch_xla.  

Some of the issues I ran into while adding this support:
- with torch_xla on GPUs, dist.is_initialized()returns false
    - Everything is handled by ENV vars
- composer.utils.dist.initialize_dist is called twice in YAHP flow.
    - Which make torch_xla thing we have 2x the number of GPUS
- There are some env vars that we need to setup before forking
- xm.* functions cannot be called until torch_xla env. is init properly.
    - We call them before init in constructing device object.
- broadcast doesn't exist in current release of torch_xla
  - Implemented with all_reduce. 

Test: 

`composer -n 2 examples/run_composer_trainer.py -f composer/yamls/models/resnet9_cifar10.yaml --train_dataset.cifar10.datadir /tmp/cifar --train_dataset.cifar10.download true --val_dataset.cifar10.datadir /tmp/cifar --val_dataset.cifar10.download true --max_duration 2ep --train_subset_num_batches 2 --device 'tpu'`

### CODE REVIEW

**Comment by dskhudia** [comment]
@hanlint : device_tpu should really be renamed to device_xla.

**Comment by mvpatel2000** [comment]
> composer.utils.dist.initialize_dist is called twice in YAHP flow.

is this function not idempotent? why does calling it more than once matter? (asking bc we'll hit this soon if people start calling it outside trainer for entrypoints)

File: composer/utils/misc.py [hunk 1]
```diff
@@ -62,3 +65,14 @@ def model_eval_mode(model: torch.nn.Module):
...
+
+def is_torch_xla_installed() -> bool:
+    """Whether torch_xla is installed in the current environment."""
+    try:
+        import torch_xla.core.xla_model as xm
+        del xm
+    except ImportError:
+        return False
+    else:
+        return True
```
Comment (mvpatel2000) [inline composer/utils/misc.py:78]:
```suggestion
def is_torch_xla_installed() -> bool:
    """Whether torch_xla is installed in the current environment."""
    try:
        import torch_xla.core.xla_model as xm
        del xm
    except ModuleNotFoundError:
        return False
    else:
        return True
```
nit: if there's an error where module is found but it fails at some other step of importing, do we want to raise this to user? my instinct is yes

File: composer/trainer/ddp.py [hunk 2]
```diff
@@ -104,7 +109,11 @@ def prepare_ddp_module(module: torch.nn.Module, find_unused_parameters: bool) ->
             to find parameters to not expect gradients for. This is useful if there are some
             parameters in the model that are not being trained.
     """
-    if dist.is_available() and dist.is_initialized():
+    # No ddp wrapping for torch_xla devices
+    if is_torch_xla_installed() and xenv.LOCAL_WORKER in os.environ:
+        return module
+
+    if (dist.is_available() and dist.is_initialized()):
```
Comment (mvpatel2000) [inline composer/trainer/ddp.py:116]: minor nit
```suggestion
    if dist.is_available() and dist.is_initialized():
```

File: composer/trainer/devices/device_tpu.py [hunk 3]
```diff
@@ -27,9 +28,8 @@ class DeviceTPU(Device):
     """
 
     def __init__(self):
-        import torch_xla.core.xla_model as xm
-
-        self._device = xm.xla_device()
+        # Can't use xm.xla_device() api here!
+        self._device = torch.device(f'xla:0')
```
Comment (mvpatel2000) [inline composer/trainer/devices/device_tpu.py:32]: Why can't we use it here?

File: composer/utils/dist.py [hunk 4]
```diff
@@ -256,6 +294,9 @@ def broadcast_object_list(object_list: List[Any], src: int = 0) -> None:
     Returns:
         None:  ``object_list`` will be modified in-place and set to values of ``object_list`` from the ``src`` rank.
     """
+    if is_torch_xla_installed():
+        raise RuntimeError(f'broadcast_object_list is not supported with torch_xla!')
```
Comment (mvpatel2000) [inline composer/utils/dist.py:298]:
```suggestion
        raise RuntimeError(f'broadcast_object_list is not supported with torch_xla')
```
(very) minor nit, feel free to ignore. Prefer no `!` to be consistent

File: composer/utils/dist.py [hunk 5]
```diff
@@ -305,6 +346,9 @@ def all_gather_object(obj: TObj) -> List[TObj]:
     Returns:
         List[TObj]: A list of objects indexed by rank.
     """
+    if is_torch_xla_installed():
+        raise RuntimeError(f'all_gather_object is not supported with torch_xla!')
```
Comment (mvpatel2000) [inline composer/utils/dist.py:350]:
```suggestion
        raise RuntimeError(f'all_gather_object is not supported with torch_xla')
```
same as above

File: composer/utils/dist.py [hunk 6]
```diff
@@ -375,6 +419,9 @@ def initialize_dist(device: Device, timeout: datetime.timedelta):
                                f'of the current process group ({dist.get_backend()}). If you '
                                'wish to change backends, please restart the python process.')
         return
+    elif is_torch_xla_installed():
+        if xenv.LOCAL_WORKER in os.environ:
+            return
```
Comment (mvpatel2000) [inline composer/utils/dist.py:424]:
```suggestion
    elif is_torch_xla_installed() and xenv.LOCAL_WORKER in os.environ:
            return
```

**Review by mvpatel2000 — COMMENTED** [review]
Minor nits and questions. Defer to others for approval because less familiar with this. Can test and read through more closely if no one is familiar with it lol

**Comment by dskhudia** [comment]
> > composer.utils.dist.initialize_dist is called twice in YAHP flow.
> 
> is this function not idempotent? why does calling it more than once matter? (asking bc we'll hit this soon if people start calling it outside trainer for entrypoints)

The way it's written, calling it second time doesn't have any effect. However, for torch_xla (torch.distributed.is_initialized) always returns False for GPUs so it was going through init again and torch_xla was assuming 2x the number of devices available. It's no longer a issue as I have added a check for torch_xla as well for repeated calls.

File: composer/trainer/devices/device_tpu.py (hunk 3 above)
Comment (dskhudia) [reply composer/trainer/devices/device_tpu.py]: Updated comment. Let me know if it's still not clear.

File: composer/utils/dist.py [hunk 7]
```diff
@@ -227,6 +256,15 @@ def broadcast(tensor: torch.Tensor, src: int) -> None:
             and tensor to be used to save received data otherwise.
         src (int): Source rank
     """
+    if is_torch_xla_installed() and xm.is_xla_tensor(tensor):
```
Comment (hanlint) [inline composer/utils/dist.py:259]: Cleaner to break this out into its own function, e.g `broadcast_xla()`

**Review by hanlint — COMMENTED** [review]
Looks good, but general comment is to break out the xla specific functionality into its own method (e.g. xla_all_reduce, xla_broadcast) for for modularity.

**Comment by dskhudia** [comment]
> Looks good, but general comment is to break out the xla specific functionality into its own method (e.g. xla_all_reduce, xla_broadcast) for for modularity.

will do

**Comment by hanlint** [comment]
> In most cases except broadcast it will be thin wrapper. Did you mean I should do it for every function or just where we have some extra logic?

Yeah, I think it would be cleaner if all functions were like:
```
def broadcast():
    if xla:
        _xla_broadcast(..)
    else:
         _dist_broadcast(..)  # or dist.broadcast if it really is just a one-liner
```

it's more verbose, but easier to read. The helper functions can be private, so no docstrings needed.

**Comment by bandish-shah** [comment]
Closing out, we'll re-visit in the future.

**Pull request closed without merging**